### PR TITLE
Implement "public import"

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR implements "public import"

Resolves https://github.com/Cray/chapel-private/issues/843

Changes:
- Adjust parser
- Add `ResolveScope::lookupPublicImports`
- Use this helper in `resolveModuleCall` in `scopeResolve`
- Adds tests
  - Some new tests for error messages
  - Some tests from https://github.com/Cray/chapel-private/issues/843
  - Copy of bunch of tests in `test/visibility/import` that are adjusted for
    `public import`

Test:
- [x] standard
- [x] gasnet

### CODE REVIEW

**Comment by e-kayrakli** [comment]
@lydia-duncan, this is currently a draft, because it can still wait for your PR
as we planned. I am still requesting review but you can hold until your work
goes in and this is updated. Up to you.

The biggest question code-wise is whether we need `lookupPublicImports` helper,
or would it be better to merge that logic into `lookupNameLocally`.

File: compiler/passes/ResolveScope.cpp [hunk 1]
```diff
@@ -830,6 +831,30 @@ Symbol* ResolveScope::lookupNameLocally(const char* name, bool isUse) const {
   return retval;
 }
 
+Symbol* ResolveScope::lookupPublicImports(const char* name) const {
+  UseImportList useImportList = mUseImportList;
+  Symbol *retval = NULL;
+
+  for (size_t i = 0; i < useImportList.size(); i++) {
```
Comment (lydia-duncan) [inline compiler/passes/ResolveScope.cpp]: Michael pointed out to me in my last PR that we can use the `for_vector_allowing_0s` macro in places like this

File: compiler/passes/scopeResolve.cpp [hunk 2]
```diff
@@ -1405,8 +1405,15 @@ static void resolveModuleCall(CallExpr* call) {
 
         currModule->moduleUseAdd(mod);
 
+        Symbol* sym;
+
         // First, try regular scope resolution
-        Symbol* sym = scope->lookupNameLocally(mbrName);
+        sym = scope->lookupNameLocally(mbrName);
```
Comment (lydia-duncan) [inline compiler/passes/scopeResolve.cpp]: Why did you break this onto a separate line?

File: test/visibility/import/public/hideHierarchy/MyLibrary.chpl [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module MyLibrary {
+  public import Unimportant.Important1;
+  public import Unimportant.Important2;
```
Comment (lydia-duncan) [inline test/visibility/import/public/hideHierarchy/MyLibrary.chpl]: Did you end up going back and looking again at the version where Unimportant is also nested?

Comment (e-kayrakli) [reply test/visibility/import/public/hideHierarchy/MyLibrary.chpl]: [this commit](https://github.com/chapel-lang/chapel/pull/15192/commits/219e505641476f33ca21485398e156d3689d180d) does that

File: compiler/passes/scopeResolve.cpp (hunk 2 above)
Comment (e-kayrakli) [reply compiler/passes/scopeResolve.cpp]: I.... don't know?

File: test/visibility/import/public/enablesUnqualified/importAndImport.chpl [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// Tests behavior when a module is imported from two different paths
+module A {
+  var x: int = 4;
+  proc foo() {
+    writeln("In A.foo()");
+  }
+}
+module B {
+  // Note: this import is private and thus does not impact the symbols available
```
Comment (lydia-duncan) [inline test/visibility/import/public/enablesUnqualified/importAndImport.chpl]: You'll want to update these comments :)

Comment (e-kayrakli) [reply test/visibility/import/public/enablesUnqualified/importAndImport.chpl]: Maintaining comments is more difficult than maintaining code...

Comment (lydia-duncan) [reply test/visibility/import/public/enablesUnqualified/importAndImport.chpl]: Ain't that the truth XD

**Pull request merged**